airbyte-cdk/python/airbyte_cdk/sources/declarative/migrations/legacy_to_per_partition_state_migration.py

```diff
@@ -0,0 +1,86 @@
+from typing import Any, Mapping
+
+from airbyte_cdk.sources.declarative.migrations.state_migration import StateMigration
+
+from airbyte_cdk.sources.declarative.interpolation.interpolated_string import InterpolatedString
+
+from airbyte_cdk.sources.declarative.models import SubstreamPartitionRouter
+
+from airbyte_cdk.sources.declarative.models import DatetimeBasedCursor
+
+
+class LegacyToPerPartitionStateMigration(StateMigration):
+
+    def __init__(self, partition_router: SubstreamPartitionRouter, cursor: DatetimeBasedCursor, config: Mapping[str, Any], parameters: Mapping[str, Any]):
+        self._partition_router = partition_router
+        self._cursor = cursor
+        self._config = config
+        self._parameters = parameters
+
+    def should_migrate(self, stream_state: Mapping[str, Any]) -> bool:
```

Comment:
I believe a long-term solution should use versioning to decide whether the state should be migrated (eg "here is a migrator to go from 1.0.0 to 2.0.0"). We don't really follow semver today so I'm going with a more hacky solution.

Since the `StateMigration` interface is pretty flexible, we can still implement versioned-based migrators in the future without the need for a breaking change